### PR TITLE
Add YOLO semantic segmentation support

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- add YOLO semantic segmentation as a sixth task in the Swift package and app UI
- add Core ML semantic output postprocessing, dense class map results, and mask overlay rendering
- add semantic model registry entries, test download support, docs, and focused unit coverage

## Production readiness
- uploaded YOLO26 semantic Core ML release assets to `v8.3.0`, which is the production tag used by `RemoteModels.swift`
- verified release asset metadata for `yolo26{n,s,m,l,x}-sem.mlpackage.zip` with matching SHA-256 digests
- verified `yolo26n-sem.mlpackage` compiles with `xcrun coremlcompiler`
- confirmed CI downloads models before resolving Swift package dependencies

## Tests
- `git diff --check`
- `xcodebuild -scheme YOLO -destination "platform=iOS Simulator,name=iPhone 17,OS=26.4.1" test`
- `xcodebuild -project YOLOiOSApp/YOLOiOSApp.xcodeproj -scheme YOLOiOSApp -destination "platform=iOS Simulator,name=iPhone 17,OS=26.4.1" build`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR adds **semantic segmentation** support to the Ultralytics YOLO iOS app and Swift package, while also improving result rendering, model handling, and documentation 📱✨

### 📊 Key Changes
- Added a new `YOLOTask.semantic` task across the SDK, app UI, examples, and tests 🧠
- Introduced a new `SemanticSegmenter` predictor to process dense per-pixel class maps from YOLO26 semantic segmentation models 🎨
- Extended `YOLOResult` with a new `semanticMask` output type for semantic segmentation results
- Added semantic segmentation visualization, including overlay rendering on images and live camera views 👀
- Included the new bundled model resource `yolo26n-sem.mlpackage` and updated scripts to export/download semantic models automatically 📦
- Updated the app UI to expose a new **Semantic** task option and made **Detect** the default task
- Simplified and unified label styling for detections, classifications, and OBB results for more consistent visuals ✨
- Refactored OBB rendering to reuse the standard bounding-box display path, reducing custom rendering complexity 🔧
- Updated README files, examples, and test docs to clearly distinguish **instance segmentation** vs **semantic segmentation** and to point users to the correct iOS app release assets 📚
- Added test coverage for semantic segmentation postprocessing, task enums, and result structures ✅
- Bumped the app marketing version from `8.8.6` to `8.8.7` 🚀

### 🎯 Purpose & Impact
- Users can now run **semantic segmentation** models directly in the iOS app and Swift package, expanding supported vision tasks beyond detection, instance segmentation, classification, pose, and OBB 🎯
- Developers get a cleaner and more complete API, with semantic outputs integrated into the same prediction flow as other YOLO tasks 🛠️
- Live and static-image rendering becomes more polished and consistent, especially for labels and OBB visualizations 🎨
- Model management is easier thanks to updated export/download scripts and clearer release/download instructions 📥
- Documentation is more accurate for both beginners and advanced users, reducing confusion around segmentation model types 📖
- There is a small compatibility consideration: apps using exhaustive `switch` statements over `YOLOTask` must now handle the new `.semantic` case ⚠️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolo-ios-app` 🚀 PR! This is an automated message to help streamline review—an engineer will assist you soon. Please review the checklist below to help ensure a smooth integration, especially for this new semantic segmentation support ✨

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolo-ios-app/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolo-ios-app` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: Sources/YOLO/SemanticSegmenter.swift [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  private func firstFeatureArray(_ request: VNRequest) -> MLMultiArray? {
+    (request.results as? [VNCoreMLFeatureValueObservation])?.first?.featureValue.multiArrayValue
+  }
+
+  func postProcessSemantic(_ logits: MLMultiArray) -> SemanticMask? {
+    let shape = logits.shape.map { $0.intValue }
+    let strides = logits.strides.map { $0.intValue }
+    guard shape.count == 4, shape[0] == 1 else {
```
Comment (UltralyticsAssistant) [inline Sources/YOLO/SemanticSegmenter.swift:53]: ⚠️ **HIGH**: This implementation only accepts 4D `[1, C, H, W]`/`[1, H, W, C]` outputs, but the PR description and docs describe Core ML semantic models as producing a dense class-map output. If the exported model returns an already-argmaxed class map (commonly `[1, H, W]` or `[H, W]`), this guard will reject valid predictions and all semantic results will come back empty. Please handle the class-map shape(s) here in addition to per-class logits.

File: YOLOiOSApp/README.md [hunk 2]
```diff
@@ -151,7 +151,8 @@ The test suite is designed to run with or without the actual Core ML model files
 To execute the complete test suite (with `SKIP_MODEL_TESTS = false`), include the following **INT8 quantized Core ML models** in your project:
 
 - **Detection:** `yolo26n.mlpackage` (place in `YOLO/DetectModels`)
-- **Segmentation:** `yolo26n-seg.mlpackage` (place in `YOLO/SegmentModels`)
+- **Instance Segmentation:** `yolo26n-seg.mlpackage` (place in `YOLO/SegmentModels`)
+- **Semantic Segmentation:** `yolo26n-sem.mlpackage` (place in `YOLO/SemanticModels`)
```
Comment (UltralyticsAssistant) [inline YOLOiOSApp/README.md]: 📝 **LOW**: The destination path here doesn't match the app's actual semantic model folder added in this PR (`Models/Semantic`). Following these instructions will place the model where the app won't discover it.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Semantic segmentation support is wired through the package, app, and tests cleanly overall. The main issue is that the new post-processing currently appears to assume logits-only 4D outputs, which conflicts with the documented dense class-map export and could cause semantic inference to fail for valid Core ML outputs. There is also a README path mismatch for where to place semantic models in the iOS app.

💬 Posted 2 inline comments

File: Sources/YOLO/SemanticSegmenter.swift (hunk 1 above)
Comment (glenn-jocher) [reply Sources/YOLO/SemanticSegmenter.swift:53]: This is fixed at the source in 0f47d9b by correcting the docs wording that caused the mismatch. The current YOLO26 semantic CoreML export emits logits (`[1, C, H, W]`, observed from the exported package), so the Swift postprocess intentionally stays logits-only instead of adding an unused class-map compatibility path.

File: YOLOiOSApp/README.md (hunk 2 above)
Comment (glenn-jocher) [reply YOLOiOSApp/README.md]: Kept the existing README placement convention as `YOLO/SemanticModels` for consistency with the surrounding task entries. The inconsistent local-only change to `YOLOiOSApp/Models/Semantic` was not pushed.

**Comment by glenn-jocher** [comment]
Correction on the README placement thread: the committed docs now use the runtime `Models/{TaskName}` folders, including `Models/Semantic`. The older `YOLO/SemanticModels` wording is stale and does not describe the current PR diff.

**Comment by glenn-jocher** [comment]
Pushed a rendering dedup pass in `59bd3a9`. Detect, pose/segment boxes, and OBB now share the existing `BoundingBoxView` label/box styling. OBB no longer has its own live `OBBRenderer`; the shared box view accepts an optional rotation angle. Static OBB image rendering now uses the same label text style and rounded label background.\n\nValidation: `xcodebuild -scheme YOLO -destination generic/platform=iOS Simulator build`; `xcodebuild -scheme YOLO -destination platform=iOS Simulator,name=iPhone 17 test`; `git diff --check`.

**Comment by glenn-jocher** [comment]
Updated classify rendering to use the same compact detection label style instead of the oversized centered label. This keeps classify visually aligned with detect/segment/pose/OBB and trims the static-image label path as well.\n\nValidation:\n- `git diff --check`\n- `xcodebuild -scheme YOLO -destination 'platform=iOS Simulator,name=iPhone 17' test`\n- post-rebase `xcodebuild -scheme YOLO -destination 'generic/platform=iOS Simulator' build`

**Comment by glenn-jocher** [comment]
Follow-up fix: classify labels now keep the compact shared label style but render centered again to avoid colliding with the top controls. The shared label alpha also has a 35% visible floor, so classify remains visible even for low-confidence outputs.\n\nValidation:\n- `git diff --check`\n- `xcodebuild -scheme YOLO -destination 'platform=iOS Simulator,name=iPhone 17' test`\n- hosted PR checks are green on `c48f062`

**Comment by UltralyticsAssistant** [comment]
Merged — fantastic work, @glenn-jocher! 🚀

As Leonardo da Vinci said, *“Simplicity is the ultimate sophistication.”* This PR is a great example of that in action: adding semantic segmentation support across the iOS app and Swift package while also simplifying rendering paths, improving model handling, expanding tests, and sharpening the docs.

Bringing `YOLOTask.semantic`, `SemanticSegmenter`, and `semanticMask` into the same streamlined workflow makes the app more capable and the developer experience more complete. The cleaner visuals, clearer segmentation guidance, and updated YOLO26 model support make this a meaningful step forward for everyone building with the Ultralytics iOS stack.

Thanks for pushing the platform forward with such thoughtful, high-impact improvements 📱✨
